### PR TITLE
[SPARK-24433][K8S] Initial R Bindings for SparkR on K8s

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

Introducing R Bindings for Spark R on K8s

- [x] Running SparkR Job

## How was this patch tested?

This patch was tested with 

- [x] Unit Tests
- [x] Integration Tests

## Example:

Commands to run example spark job:
1. `dev/make-distribution.sh --pip --r --tgz -Psparkr -Phadoop-2.7 -Pkubernetes`
2. `bin/docker-image-tool.sh -m -t testing build`
3. 
```
bin/spark-submit \
    --master k8s://https://192.168.64.33:8443 \
    --deploy-mode cluster \
    --name spark-r \
    --conf spark.executor.instances=1 \
    --conf spark.kubernetes.container.image=spark-r:testing \
    local:///opt/spark/examples/src/main/r/dataframe.R
```

This above spark-submit command works given the distribution. (Will include this integration test in PR once PRB is ready).

### CODE REVIEW

**Comment by ifilonenko** [comment]
@felixcheung for initial review 
There is still a need for the integration testing environment to be setup on the PRB side. But I would like community opinion on the format of the Dockerfile.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# limitations under the License.
+#
+
+ARG base_img
+FROM $base_img
+WORKDIR /
+RUN mkdir ${SPARK_HOME}/R
+COPY R ${SPARK_HOME}/R
+
+RUN apk add --no-cache R R-dev
```
Comment (mccheah) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: Should we install a specific version of R here? Perhaps allow it to be parameterized as a build argument with some default.

Comment (ifilonenko) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: Thoughts @felixcheung ?

**Comment by felixcheung** [comment]
We have a min supported version of R
https://github.com/apache/spark/blob/master/R/pkg/DESCRIPTION

But beyond that we haven’t seen much problems when new releases of R are coming up.

It might be helpful to make it configurable if any user has any particular need with R and/or R packages versio

**Comment by ifilonenko** [comment]
@felixcheung @mccheah by that logic.. should then both Python and R images be version configurable? This is all handled in the `bin/docker-image-tool.sh` ? Aren't those configurations just way to hard to deal with (i.e. make the build script very cumbersome with all the build-args... why can't they just create their own images with unique versions? i.e. 
```
ARG VERSION
apk-add R-base:$VERSION R-dev:$VERSION
```

**Comment by felixcheung** [comment]
IMO it's fine we have one version supported in the image and stick with that.
the tricky thing is having maintainers to keep updating/testing the newer versions in the images (we have history of not able to keep up)

would it be possible for the integration test to build the image running `docker-image-tool.sh -m -t testing build
`, and then run the integration test with it?

**Comment by ifilonenko** [comment]
@felixcheung The integration tests already run `docker-image-tool.sh -t testing build` which build the `spark-base`, `spark-py`, and `spark-r` images. At which point integration tests would be run from the `KubernetesSuite`. As such, is this okay as an initial PR to get SparkR support, (in its current state) after integration tests are included in this PR?

**Comment by felixcheung** [comment]
yap

**Comment by shaneknapp** [comment]
just wanted to briefly chime in:

TL;DR:  this build will fail until the PRB is running on our ubuntu build nodes.

we are currently blocked from testing this stuff w/the current iteration of the PRB on our jenkins for a couple of reasons:

1) the current pull request builder runs only on our old, ancient and crufty centos 6.9 hosts
2) these hosts have old, ancient and crufty installations of R (3.1.1) and associated libraries
3) trying to re-create (2) on our shiny, new ubuntu 16.04 boxen has, so far, been an exercise in futility.  :\
4) this is why we currently have a limited number of ubuntu boxes, and also why things like testing for spark/k8s integration requires an addition build *on top* of the standard PRB build to make things work.

i'm working now w/some folks at databricks (@weshsiao) to get this under control enough so that we can get things moving to the ubuntu hosts ASAP.

**Comment by ifilonenko** [comment]
All tests pass locally. Waiting on PRB to be configured. 
```
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run SparkR on simple dataframe.R example
```

**Comment by ifilonenko** [comment]
Ready for merge upon PRB passing the tests.

File: bin/docker-image-tool.sh [hunk 2]
```diff
@@ -79,11 +80,16 @@ function build {
   docker build $NOCACHEARG "${BINDING_BUILD_ARGS[@]}" \
     -t $(image_ref spark-py) \
     -f "$PYDOCKERFILE" .
+
+    docker build "${BINDING_BUILD_ARGS[@]}" \
```
Comment (felixcheung) [inline bin/docker-image-tool.sh]: nit: this doesn't align with the pattern above

Comment (felixcheung) [reply bin/docker-image-tool.sh]: is should `$NOCACHEARG` too?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala [hunk 3]
```diff
@@ -71,8 +75,9 @@ private[spark] class KubernetesDriverBuilder(
         case JavaMainAppResource(_) =>
           provideJavaStep(kubernetesConf)
         case PythonMainAppResource(_) =>
-          providePythonStep(kubernetesConf)}
-      .getOrElse(provideJavaStep(kubernetesConf))
+          providePythonStep(kubernetesConf)
+        case RMainAppResource(_) =>
+          provideRStep(kubernetesConf)}.getOrElse(provideJavaStep(kubernetesConf))
```
Comment (felixcheung) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: I'd move `}.getOrElse(provideJavaStep(kubernetesConf))` to the next line for clarity?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (felixcheung) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: add `rm -r /root/.cache` like in python?

**Review by felixcheung — COMMENTED** [review]
cool, thanks, minor comments

**Comment by felixcheung** [comment]
looks like test failed

```
- Run SparkR on simple dataframe.R example *** FAILED ***
  The code passed to eventually never returned normally. Attempted 70 times over 2.0071955115 minutes. Last failure message: Failure executing: GET at: https://192.168.39.28:8443/api/v1/namespaces/3346b714f192421eaae8d115154cec0e/pods/spark-test-app-da59d78dbe144e7e84581429199be99a/log?pretty=false. Message: container "spark-kubernetes-driver" in pod "spark-test-app-da59d78dbe144e7e84581429199be99a" is waiting to start: image can't be pulled. Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=container "spark-kubernetes-driver" in pod "spark-test-app-da59d78dbe144e7e84581429199be99a" is waiting to start: image can't be pulled, metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).. (KubernetesSuite.scala:221)
```

**Comment by ifilonenko** [comment]
@felixcheung the PRB isn’t currently configured with capability to build —r distributions. Waiting on @shaneknapp, which is why it is unable to build the image. With an R supported environment, it passes all tests.

**Comment by felixcheung** [comment]
I see. you are referring to https://github.com/apache/spark/pull/21584#issuecomment-406091994
got it.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile (hunk 4 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: This is only for Python packaging. R does not have `/root/.cache` when created by alpine.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala (hunk 3 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: done

**Comment by felixcheung** [comment]
SPARK-24908 went in, it should unblock R?

**Comment by shaneknapp** [comment]
@felixcheung -- SPARK-24908 is just one of many layers of the onion that i'm having to peel away to get the GHPRB ported to ubuntu.  we're close to getting it to reliably pass, but not quite there yet.  there's another potential bug w/some scala unit tests that i will be creating an issue for today.

once i get the GHPRB working reliably enough for us to deem it safe enough to move to ubuntu, we'll need to start wiping the existing centos workers and reinstalling them w/ubuntu.  this will definitely take time...  and i'm guessing at the minimum a couple of weeks for this to happen.

i'm doing my best to get this done but there's no way that we'll get things finished before 2.4 is cut.

**Comment by ifilonenko** [comment]
retest this please

File: bin/docker-image-tool.sh [hunk 5]
```diff
@@ -97,12 +103,13 @@ Commands:
...
 
 Options:
-  -f file     Dockerfile to build for JVM based Jobs. By default builds the Dockerfile shipped with Spark.
-  -p file     Dockerfile with Python baked in. By default builds the Dockerfile shipped with Spark.
-  -r repo     Repository address.
-  -t tag      Tag to apply to the built image, or to identify the image to be pushed.
-  -m          Use minikube's Docker daemon.
-  -n          Build docker image with --no-cache
+  -f file               Dockerfile to build for JVM based Jobs. By default builds the Dockerfile shipped with Spark.
+  -p file               Dockerfile with Python baked in. By default builds the Dockerfile shipped with Spark.
```
Comment (holdenk) [inline bin/docker-image-tool.sh]: nit: Maybe for -p & -R clarify that it build's the Python or R docker file shipped with Spark since we are going to be shipping multiple ones and if someone goes and looks at the default one they might get confused about how this builds.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 6]
```diff
@@ -205,7 +218,7 @@ private[spark] object Config extends Logging {
       .createWithDefault(0.1)
 
   val PYSPARK_MAJOR_PYTHON_VERSION =
-    ConfigBuilder("spark.kubernetes.pyspark.pythonversion")
+    ConfigBuilder("spark.kubernetes.pyspark.pythonVersion")
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: nit / curiousity: why are we changing this?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: I think this could make sense to support, the same could be true for Python as well since minor version mismatch can lead to very bad behaviour. It _might_ make sense to do in a follow up PR since we probably can't install all the versions a user might want from apk?

**Review by holdenk — COMMENTED** [review]
Did a quick first pass with some comments.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 6 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: idiomatic for configs to have camelCase.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile (hunk 1 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: Agreed. Minor versions are something we *might* want to support in the future. Let's look to include this in a follow-up PR if there is enough desire in custom version arguments to be passed as build_args to the `bin/docker-image-tool.sh`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 6 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: K

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: We _might_ want to make that (small) change in a separate PR just so that if this slips and doesn't make the cut for 2.4 we aren't trying to change a config string later.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: hmm good call.... I might do that then

**Comment by mccheah** [comment]
Let's merge at the end of the day pacific time (~5PM-ish) on Friday, August 17, pending any additional feedback on the mailing list thread discussing the subject of including this in 2.4.

**Comment by felixcheung** [comment]
LGTM

**Comment by shaneknapp** [comment]
LGTM

**Comment by mccheah** [comment]
Ok I am merging this to master now. Thanks for the work on this!

**Comment by mccheah** [comment]
I filed https://issues.apache.org/jira/browse/SPARK-25152 for the integration tests.
